tools/themis/src/utils.rs

```diff
@@ -0,0 +1,135 @@
+        chk!(@[failed] [$workspace]: { $($rules)+ });
+        !failed
+    }};
+    (@[$failed:expr] [$workspace:ident]: {# $rule:expr $(, $( $($rest:tt)+ )? )?}) => {
+        let _ = $rule;
+        $( $( chk!(@[$failed] [$workspace]: { $($rest)+ }) )? )?
+    };
+    (@[$failed:expr] [$workspace:ident]: {$rule:expr $(, $( $($rest:tt)+ )? )?}) => {
+        $failed |= $crate::utils::check_and_report($rule(&$workspace), &$workspace)?;
+        $( $( chk!(@[$failed] [$workspace]: { $($rest)+ }) )? )?
+    };
+}
+
+pub fn check_and_report<'a>(
+    outcome: Result<(), Error<'a>>,
+    workspace: &Workspace,
+) -> anyhow::Result<bool> {
+    match outcome {
+        Err(Error::RuntimeError(err)) => Err(err),
+        Err(Error::OutcomeError { msg, expected, outliers }) => {
+            let header = format!(
+                "{c_heading}(i) {}:{c_none}{}",
+                msg,
+                expected.map_or("".to_string(), |Expected { value, reason }| format!(
+                    " [expected: {c_expected}{}{c_none}{}]",
+                    value,
+                    reason.map_or("".to_string(), |reason| format!(", {}", reason)),
+                    c_expected = style::fg(style::Color::Color256(35))
+                        + &style::bg(style::Color::Gray { shade: 3 })
+                        + style::bold(),
+                    c_none = style::reset()
+                )),
+                c_heading = style::fg(style::Color::Color256(172)) + style::bold(),
+                c_none = style::reset()
+            );
+            let report = outliers.into_iter().fold(
+                header,
+                |mut acc, PackageOutcome { pkg: Package { parsed: pkg, .. }, value }| {
+                    acc.push_str(&format!(
+                        "\n \u{2022} {c_name}{}{c_none} v{} {c_path}({}){c_none}{}",
+                        pkg.name,
+                        pkg.version,
+                        pkg.manifest_path.strip_prefix(&workspace.root).unwrap(),
+                        value.map_or("".to_string(), |v| format!(
+                            " [found: {c_found}{}{c_none}]",
+                            v,
+                            c_found = style::fg(style::Color::White)
+                                + &style::bg(style::Color::Red)
+                                + style::bold(),
+                            c_none = style::reset()
+                        )),
+                        c_name = style::fg(style::Color::Color256(39)) + style::bold(),
+                        c_path = style::fg(style::Color::Gray { shade: 12 }),
+                        c_none = style::reset(),
+                    ));
+                    acc
+                },
+            );
+            eprintln!("{}", report);
+            Ok(true)
+        }
+        _ => Ok(false),
+    }
+}
+
+pub fn is_publishable(pkg: &Package) -> bool {
+    !matches!(pkg.raw["package"].get("publish"), Some(toml::Value::Boolean(false)))
```

Comment:
No need to reach into `raw` to get this, this is exposed via cargo metadata: 

https://docs.rs/cargo_metadata/latest/cargo_metadata/struct.Package.html#structfield.publish

Reply:
Well, I thought so too. Until I realized, according to the doc:

> Publishing is unrestricted if `None`, and forbidden if the `Vec` is empty.

This doesn't disambiguate between omitting the field and explicitly specifying `publish = true`. Whereas, in our case, we need to fail if the field is omitted.